### PR TITLE
feat: hermetic library generation for showcase

### PR DESCRIPTION
**Pull request by diegomarquezp**
This script follows from https://github.com/googleapis/sdk-platform-java/pull/1781

This PR provides the `generate_library.sh` and it is tested against the `sdk-platform-java/showcase` folder. The current output of the script exactly matches the contents of gapic showcase _except for the comments added in the generated grpc library_.

A few remarks:
* No readme was added to avoid duplication of work in https://github.com/googleapis/sdk-platform-java/pull/1916
* The [proto folder in showcase](https://github.com/googleapis/sdk-platform-java/tree/cc6746ef1f8ab1256028395f623ed27155564ee3/showcase/proto-gapic-showcase-v1beta1/src/main) contains an irregular path, different from the one computed in monorepo/googleapis-gen libraries. The location of the protos looks like `proto-gapic-showcase-v1beta1/src/main/proto/external/com_google_gapic_showcase/schema/google/showcase
/v1beta1`, while if it was a monorepo folder it would look like `proto-gapic-showcase-v1beta1/src/main/proto/schema/google/showcase
/v1beta1`. The `schema` folder is the first folder in the path from the `gapic-showcase` repo to the [folder location of the showcase protos](https://github.com/googleapis/gapic-showcase/tree/main/schema/google/showcase/v1beta1), just like it would be with googleapis-based libraries. This has a [temporary workaround in the generation script](https://github.com/googleapis/sdk-platform-java/blob/93f55ae6211e7a4bbb386b064f8153b179b17272/library_generation/generate_library.sh#L198C32-L198C32).
* There are a few other parts that depend on the path from source repository root (e.g. [googleapis](https://github.com/googleapis/googleapis), [gapic-showcase](https://github.com/googleapis/gapic-showcase/tree/main)) to the location of the protos. This was solved by adding an optional input `--proto-path` (as opposed to `--proto-location` which points to the folder in the filesystem containing the protos). `--proto-path` must be a path in the style of `google/cloud/service/v1beta1`, for example. If it's not provided a [util function](https://github.com/googleapis/sdk-platform-java/blob/93f55ae6211e7a4bbb386b064f8153b179b17272/library_generation/util.sh#L129-L151) will heuristically determine this path by calculating the path from the git repository root (i.e. folder containing `.git` folder) to the inner folder containing the protos.
Examples of dependent locations:
  * [Comments in grpc library](https://github.com/googleapis/google-cloud-java/blob/0b28660f6424a9c26e8c9befe3a1f0660c317217/java-channel/grpc-google-cloud-channel-v1/src/main/java/com/google/cloud/channel/v1/CloudChannelReportsServiceGrpc.java#L31)
  * [Comments in proto library](https://github.com/googleapis/google-cloud-java/blob/0b28660f6424a9c26e8c9befe3a1f0660c317217/java-channel/proto-google-cloud-channel-v1/src/main/java/com/google/cloud/channel/v1/ActivateEntitlementRequest.java#L17)
* The showcase generated grpc library does not contain comments. If its built via `bazel build --experimental_proto_descriptor_sets_include_source_info //showcase:google-cloud-showcase-v1beta1-java` the comments will be available (and we would have a perfect match)
* Post processing was originally in this PR but it was removed in https://github.com/googleapis/sdk-platform-java/pull/1906/commits/66afe996a33cec6ac79894069829e70129e83a63. It will be used for a later stage of the project.
* The name of the gapic, grpc and proto files follow a similar naming pattern. For example `proto-[gapic|google-cloud]-showcase-v1beta1`. The script currently extracts the service name (e.g. `showcase`) and service version (e.g. `v1beta1`) from the `$PROTO_PATH` variable (see point above on `--proto-path`).
* A github action checks out `gapic-showcase` and generates the libraries from the protos in `schema/google/showcase/v1beta1` into the `sdk-platform-java/showcase` and produces a diff

### CODE REVIEW

File: library_generation/generate_library.sh [hunk 1]
```diff
@@ -0,0 +1,298 @@
+#!/usr/bin/env bash
+
+set -ex
+GOOGLEAPIS_COMMIT=$1
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/generate_library.sh:6]: note: commit is not necessary. We just need a path to the folder containing the protos

**Comment by diegomarquezp** [comment]
To work on showcase test in this PR

**Comment by diegomarquezp** [comment]
To solve difference in grpc version annotation

File: library_generation/generate_library.sh
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/generate_library.sh:1]: done

File: library_generation/post-processing/consolidate_config.sh
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/post-processing/consolidate_config.sh:1]: removed as post-processing was left for a later stage

File: library_generation/generate_library.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/generate_library.sh]: changed. Added notes on extra input

File: library_generation/generate_library.sh [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+# copy repo metadata to destination library folder
+WORKSPACE=$LIBRARY_GEN_OUT/workspace
+mkdir $WORKSPACE
+cp $REPO_METADATA_PATH $WORKSPACE/.repo-metadata.json
+
+# call owl-bot-copy
+OWLBOT_STAGING_FOLDER="$WORKSPACE/owl-bot-staging"
+OWLBOT_IMAGE=gcr.io/cloud-devrel-public-resources/owlbot-java@sha256:$OWLBOT_SHA
+
+# render owlbot yaml template
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/generate_library.sh]: Post processing for later stage

File: gapic-generator-java/generate_sdk_library.sh
Comment (JoeWang1127) [inline gapic-generator-java/generate_sdk_library.sh:1]: I think this file should be removed as well?

File: .github/workflows/verify_library_generation.yaml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        REST_NUMERIC_ENUMS=false # true or false
+        INCLUDE_SAMPLES=false # true or false
+        bash gapic-generator-java/library-generation/generate_library.sh \
+            --proto-location $PROTO_LOCATION \
+            --destination-location $DESTINATION_PATH \
+            --gapic-generator-java-version $GAPIC_GENERATOR_VERSION \
+            --protobuf-version $PROTOBUF_VERSION \
+            --grpc-version $GRPC_VERSION \
+            --transport $TRANSPORT
+    - name: compare generation result
```
Comment (suztomo) [inline .github/workflows/verify_library_generation.yaml:56]:
```suggestion
    - name: compare generation result (This fails if any diff exists)
```

File: .github/workflows/verify_library_generation.yaml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        sudo apt-get install libxml2-utils
+        echo "GENERATOR_VERSION=$(xmllint --xpath "string(//*[local-name()='version'])" gapic-generator-java-pom-parent/pom.xml)" >> $GITHUB_ENV
+        echo "PROTOBUF_VERSION=$(xmllint --xpath "string(//*[local-name()='protobuf.version'])" gapic-generator-java-pom-parent/pom.xml)" >> $$GITHUB_ENV
+        echo "GRPC_VERSION=$(xmllint --xpath "string(//*[local-name()='grpc.version'])" gapic-generator-java-pom-parent/pom.xml)" >> $$GITHUB_ENV
+    - name: clone showcase
+      run: |
+        git clone https://github.com/googleapis/gapic-showcase.git
+    - name: checkout a tag in showcase
+      run: |
+        git checkout v0.28.2
```
Comment (suztomo) [inline .github/workflows/verify_library_generation.yaml:36]: I would combine the two steps. They are tightly-coupled.

File: .github/workflows/verify_library_generation.yaml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      working-directory: gapic-showcase
+    - name: generate java-aiplatform
```
Comment (suztomo) [inline .github/workflows/verify_library_generation.yaml:38]: Typo?

File: .github/workflows/verify_library_generation.yaml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        bash gapic-generator-java/library-generation/generate_library.sh \
+            --proto-location $PROTO_LOCATION \
+            --destination-location $DESTINATION_PATH \
+            --gapic-generator-java-version $GAPIC_GENERATOR_VERSION \
+            --protobuf-version $PROTOBUF_VERSION \
+            --grpc-version $GRPC_VERSION \
+            --transport $TRANSPORT
+    - name: compare generation result
+      run: |
+        git diff showcase
```
Comment (suztomo) [inline .github/workflows/verify_library_generation.yaml:58]: Can you add a step to delete showcase Java code before generating the Java code?

**Review by suztomo — CHANGES_REQUESTED** [review]

**Review by suztomo — CHANGES_REQUESTED** [review]

File: gapic-generator-java/generate_sdk_library.sh
Comment (diegomarquezp) [reply gapic-generator-java/generate_sdk_library.sh:1]: removed

File: .github/workflows/verify_library_generation.yaml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      run: |
+        git clone https://github.com/googleapis/gapic-showcase.git
+    - name: checkout a tag in showcase
+      run: |
+        git checkout v0.28.2
+      working-directory: gapic-showcase
+    - name: generate java-aiplatform
+      run: |
+        PROTO_LOCATION=$(pwd)/gapic-showcase/schema/google/showcase/v1beta1/
+        PROTO_PATH=schema/google/showcase/v1beta1
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply .github/workflows/verify_library_generation.yaml:41]: I stopped using it to test the [path inference function](https://github.com/googleapis/sdk-platform-java/blob/93f55ae6211e7a4bbb386b064f8153b179b17272/library_generation/util.sh#L129-L151). I will remove it here as well

[the remark this answers is not in the document]
Comment (diegomarquezp) [reply .github/workflows/verify_library_generation.yaml]: removed

File: .github/workflows/verify_library_generation.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply .github/workflows/verify_library_generation.yaml:29]: Yes, I explained it's now hardcoded to follow referenced versions of `sdk-platform-java:2.24.0`

File: .github/workflows/verify_library_generation.yaml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        cache: maven
+    - name: maven install
+      run: |
+        mvn install -B -ntp -DskipTests -Dclirr.skip -Dcheckstyle.skip
+    - name: get gapic-generator-java, protobuf, grpc version
+      run: |
+        set -x
+        sudo apt-get update
+        sudo apt-get install libxml2-utils
+        echo "GENERATOR_VERSION=$(xmllint --xpath "string(//*[local-name()='version'])" gapic-generator-java-pom-parent/pom.xml)" >> $GITHUB_ENV
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply .github/workflows/verify_library_generation.yaml]: I just hardcoded the versions for now

File: .github/workflows/verify_library_generation.yaml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    - name: clone showcase
+      run: |
+        git clone https://github.com/googleapis/gapic-showcase.git
+    - name: checkout a tag in showcase
+      run: |
+        git checkout v0.28.2
+      working-directory: gapic-showcase
+    - name: generate java-aiplatform
+      run: |
+        PROTO_LOCATION=$(pwd)/gapic-showcase/schema/google/showcase/v1beta1/
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply .github/workflows/verify_library_generation.yaml]: removed redundancy

File: .github/workflows/verify_library_generation.yaml (hunk 4 above)
Comment (diegomarquezp) [reply .github/workflows/verify_library_generation.yaml]: thanks, combined

File: .github/workflows/verify_library_generation.yaml (hunk 5 above)
Comment (diegomarquezp) [reply .github/workflows/verify_library_generation.yaml]: thanks for the catch, fixed.

File: gapic-generator-java/library-generation/gapic-generator-java-wrapper [hunk 10]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply gapic-generator-java/library-generation/gapic-generator-java-wrapper:1]: Added a quick explanation of this

File: .github/workflows/verify_library_generation.yaml [hunk 11]
```diff
@@ -0,0 +1,59 @@
+on:
+  push:
+    branches:
+    - main
+  pull_request:
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply .github/workflows/verify_library_generation.yaml:5]: Added a constraint to trigger the workflow only when there is a PR in `gapic-generator-java/library-generation`

**Comment by blakeli0** [comment]
While this gives us more confidence that `generate_library.sh` can be used for generating showcase clients, I thought this PR is more about changing the current showcase generation process from bazel/maven to the script. e.g. Currently the showcase client can be regenerated by [maven command](https://github.com/googleapis/sdk-platform-java/tree/main/showcase#update-the-golden-showcase-files), we can keep the existing maven command and change the implementation to use the script, or wrap the showcase generation to another script and update README.

File: gapic-generator-java/library-generation/out [hunk 12]
```diff
@@ -0,0 +1,235 @@
+++ PROTO_LOCATION=/usr/local/google/home/diegomarquezp/Desktop/gapic-showcase/schema/google/showcase/v1beta1/
```
Comment (suztomo) [inline gapic-generator-java/library-generation/out:1]: I don't see this file useful.

**Comment by diegomarquezp** [comment]
Closing in favor of creating a PR with showcase logic against https://github.com/googleapis/sdk-platform-java/pull/1916

**Pull request closed without merging**